File: card-api/src/main/java/com/jean202/cardmizer/api/performance/PerformanceOverviewController.java
package com.jean202.cardmizer.api.performance;

import com.jean202.cardmizer.core.domain.SpendingPeriod;
import com.jean202.cardmizer.core.port.in.GetPerformanceOverviewUseCase;
import java.time.YearMonth;
import java.util.List;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/performance-overview")
public class PerformanceOverviewController {
    private final GetPerformanceOverviewUseCase getPerformanceOverviewUseCase;

    public PerformanceOverviewController(GetPerformanceOverviewUseCase getPerformanceOverviewUseCase) {
        this.getPerformanceOverviewUseCase = getPerformanceOverviewUseCase;
    }

    @GetMapping
    public List<PerformanceOverviewResponse> getOverview(
            @RequestParam
            @DateTimeFormat(pattern = "yyyy-MM")
            YearMonth yearMonth
    ) {
        return getPerformanceOverviewUseCase.getOverview(new SpendingPeriod(yearMonth)).stream()
                .map(PerformanceOverviewResponse::from)
                .toList();
    }

    public record PerformanceOverviewResponse(
            String cardId,
            String cardName,
            int priority,
            long spentAmount,
            long targetAmount,
            long remainingAmount,
            boolean achieved,
            String targetTierCode
    ) {
        static PerformanceOverviewResponse from(GetPerformanceOverviewUseCase.CardPerformanceSnapshot snapshot) {
            return new PerformanceOverviewResponse(
                    snapshot.card().id().value(),
                    snapshot.card().displayName(),
                    snapshot.card().priority(),
                    snapshot.spentAmount().amount(),
                    snapshot.targetAmount().amount(),
                    snapshot.remainingAmount().amount(),
                    snapshot.achieved(),
                    snapshot.targetTierCode()
            );
        }
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/performance/PerformanceOverviewController.kt
package com.jean202.cardmizer.api.performance

import com.jean202.cardmizer.core.domain.SpendingPeriod
import com.jean202.cardmizer.core.port.`in`.GetPerformanceOverviewUseCase
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.time.YearMonth

@RestController
@RequestMapping("/api/performance-overview")
class PerformanceOverviewController(
    private val getPerformanceOverviewUseCase: GetPerformanceOverviewUseCase,
) {
    @GetMapping
    fun getOverview(
        @RequestParam @DateTimeFormat(pattern = "yyyy-MM") yearMonth: YearMonth,
    ): List<PerformanceOverviewResponse> =
        getPerformanceOverviewUseCase.getOverview(SpendingPeriod(yearMonth))
            .map { PerformanceOverviewResponse.from(it) }

    data class PerformanceOverviewResponse(
        val cardId: String,
        val cardName: String,
        val priority: Int,
        val spentAmount: Long,
        val targetAmount: Long,
        val remainingAmount: Long,
        val achieved: Boolean,
        val targetTierCode: String,
    ) {
        companion object {
            fun from(snapshot: GetPerformanceOverviewUseCase.CardPerformanceSnapshot) =
                PerformanceOverviewResponse(
                    cardId = snapshot.card.id.value,
                    cardName = snapshot.card.displayName(),
                    priority = snapshot.card.priority,
                    spentAmount = snapshot.spentAmount.amount,
                    targetAmount = snapshot.targetAmount.amount,
                    remainingAmount = snapshot.remainingAmount.amount,
                    achieved = snapshot.achieved,
                    targetTierCode = snapshot.targetTierCode,
                )
        }
    }
}
